File: java/src/mouseracer/GameEngine.java
package mouseracer;

import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;

public class GameEngine {
    public enum GameMode {
        SPRINT, ENDURANCE, MARATHON
    }

    public interface GameListener {
        void onTick(float seconds, int totalClicks, float currentCps);
        void onClick(int totalClicks);
        void onGameOver(String taunt, int totalClicks, float seconds, float avgCps, float maxCps);
    }

    private GameMode mode;
    private final AtomicInteger clicks = new AtomicInteger(0);
    private final AtomicInteger totalClicks = new AtomicInteger(0);
    private float seconds;
    private float startSeconds;
    private float maxCps;
    private float firstClick;
    private float adjFirstClick;
    private volatile boolean isRunning;
    private ScheduledExecutorService scheduler;
    private final GameListener listener;

    public GameEngine(GameListener listener) {
        this.listener = listener;
    }

    public void start(GameMode mode) {
        if (isRunning) stop();
        
        this.mode = mode;
        this.clicks.set(0);
        this.totalClicks.set(0);
        this.maxCps = 0;
        this.firstClick = 0;
        this.isRunning = true;
        switch (mode) {
            case SPRINT:
                this.seconds = 10;
                break;
            case ENDURANCE:
            case MARATHON:
                this.seconds = 0;
                break;
        }
        this.startSeconds = this.seconds;

        scheduler = Executors.newSingleThreadScheduledExecutor();
        scheduler.scheduleAtFixedRate(() -> {
            switch (this.mode) {
                case SPRINT:
                    handleSprintTick();
                    break;
                case ENDURANCE:
                    handleEnduranceTick();
                    break;
                case MARATHON:
                    handleMarathonTick();
                    break;
            }
        }, 1, 1, TimeUnit.SECONDS);

        listener.onTick(seconds, totalClicks.get(), 0);
    }

    public void registerClick() {
        if (isRunning) {
            int currentTotal = totalClicks.incrementAndGet();
            clicks.incrementAndGet();
            listener.onClick(currentTotal);
        }
    }

    public void stop() {
        if (scheduler != null) {
            scheduler.shutdownNow();
            scheduler = null;
        }
        isRunning = false;
    }

    public boolean isRunning() {
        return isRunning;
    }

    private synchronized void handleSprintTick() {
        seconds--;
        
        int currentClicks = clicks.getAndSet(0);
        float currentCps = currentClicks;
        if (currentCps > maxCps) maxCps = currentCps;
        listener.onTick(seconds, totalClicks.get(), currentCps);

        if (seconds <= 0) {
            gameOver();
        }
    }

    private synchronized void handleEnduranceTick() {
        seconds++;
        int currentClicks = clicks.getAndSet(0);
        float currentCps = currentClicks;
        if (currentCps > maxCps) maxCps = currentCps;
        listener.onTick(seconds, totalClicks.get(), currentCps);

        if (seconds >= 100) {
            gameOver();
        }
    }

    private synchronized void handleMarathonTick() {
        seconds++;
        int currentClicks = clicks.getAndSet(0);
        float currentCps = currentClicks;
        
        if (firstClick == 0 && currentCps > 0) {
            firstClick = currentCps;
            adjFirstClick = firstClick * 0.75f;
        }

        if (currentCps > firstClick * 1.05f) {
            firstClick = currentCps;
            adjFirstClick = firstClick * 0.75f;
        }

        if (firstClick > 0 && currentCps <= adjFirstClick) {
            gameOver();
            return;
        }

        if (currentCps > maxCps) maxCps = currentCps;
        listener.onTick(seconds, totalClicks.get(), currentCps);
    }

    private void gameOver() {
        stop();
        float elapsed = (mode == GameMode.SPRINT) ? startSeconds - seconds : seconds;
        float avgCps = elapsed > 0 ? (float) totalClicks.get() / elapsed : 0;
        listener.onGameOver(getTaunt(), totalClicks.get(), elapsed, avgCps, maxCps);
    }

    private String getTaunt() {
        int total = totalClicks.get();
        if (total > 350) return "You Might Be Playing too much MouseRacer!";
        if (total > 251) return "All time high score! NOT!";
        if (total > 201) return "Unstoppable!";
        if (total > 151) return "That call stack has your name written all over it!";
        if (total > 121) return "You Have Taken The Lead";
        if (total > 101) return "you have achieved hyperclick!";
        if (total > 50) return "You make Chuck Norris Proud!";
        return "you might consider a different hobby";
    }
}

File: rust/src/game_engine.rs
use std::time::Instant;

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum GameMode {
    Sprint,
    Endurance,
    Marathon,
}


#[derive(Debug)]
pub enum GameEvent {
    Tick {
        seconds: f32,
        total_clicks: u32,
        current_cps: f32,
    },
    Click {
        total_clicks: u32,
    },
    GameOver {
        taunt: String,
        total_clicks: u32,
        seconds: f32,
        avg_cps: f32,
        max_cps: f32,
    },
}

pub struct GameEngine {
    mode: GameMode,
    clicks_this_second: u32,
    total_clicks: u32,
    seconds: f32,
    start_seconds: f32,
    max_cps: f32,
    first_click: f32,
    adj_first_click: f32,
    running: bool,
    last_tick: Option<Instant>,
}

impl GameEngine {
    pub fn new() -> Self {
        Self {
            mode: GameMode::Sprint,
            clicks_this_second: 0,
            total_clicks: 0,
            seconds: 0.0,
            start_seconds: 0.0,
            max_cps: 0.0,
            first_click: 0.0,
            adj_first_click: 0.0,
            running: false,
            last_tick: None,
        }
    }

    pub fn start(&mut self, mode: GameMode) -> GameEvent {
        self.mode = mode;
        self.clicks_this_second = 0;
        self.total_clicks = 0;
        self.max_cps = 0.0;
        self.first_click = 0.0;
        self.adj_first_click = 0.0;
        self.running = true;

        self.seconds = match mode {
            GameMode::Sprint => 10.0,
            GameMode::Endurance | GameMode::Marathon => 0.0,
        };
        self.start_seconds = self.seconds;
        self.last_tick = Some(Instant::now());

        GameEvent::Tick {
            seconds: self.seconds,
            total_clicks: 0,
            current_cps: 0.0,
        }
    }

    pub fn register_click(&mut self) -> Option<GameEvent> {
        if self.running {
            self.total_clicks += 1;
            self.clicks_this_second += 1;
            Some(GameEvent::Click {
                total_clicks: self.total_clicks,
            })
        } else {
            None
        }
    }

    pub fn is_running(&self) -> bool {
        self.running
    }

    pub fn stop(&mut self) {
        self.running = false;
        self.last_tick = None;
    }

    /// Call each frame. Returns a tick event if a second has elapsed, or a game-over event.
    pub fn tick(&mut self) -> Option<GameEvent> {
        if !self.running {
            return None;
        }

        let Some(last) = self.last_tick else {
            return None;
        };

        if last.elapsed().as_secs_f32() < 1.0 {
            return None;
        }

        self.last_tick = Some(Instant::now());

        match self.mode {
            GameMode::Sprint => self.handle_sprint_tick(),
            GameMode::Endurance => self.handle_endurance_tick(),
            GameMode::Marathon => self.handle_marathon_tick(),
        }
    }

    fn handle_sprint_tick(&mut self) -> Option<GameEvent> {
        self.seconds -= 1.0;

        let current_cps = self.clicks_this_second as f32;
        self.clicks_this_second = 0;
        if current_cps > self.max_cps {
            self.max_cps = current_cps;
        }

        if self.seconds <= 0.0 {
            return Some(self.game_over());
        }

        Some(GameEvent::Tick {
            seconds: self.seconds,
            total_clicks: self.total_clicks,
            current_cps,
        })
    }

    fn handle_endurance_tick(&mut self) -> Option<GameEvent> {
        self.seconds += 1.0;

        let current_cps = self.clicks_this_second as f32;
        self.clicks_this_second = 0;
        if current_cps > self.max_cps {
            self.max_cps = current_cps;
        }

        if self.seconds >= 100.0 {
            return Some(self.game_over());
        }

        Some(GameEvent::Tick {
            seconds: self.seconds,
            total_clicks: self.total_clicks,
            current_cps,
        })
    }

    fn handle_marathon_tick(&mut self) -> Option<GameEvent> {
        self.seconds += 1.0;

        let current_cps = self.clicks_this_second as f32;
        self.clicks_this_second = 0;

        if self.first_click == 0.0 && current_cps > 0.0 {
            self.first_click = current_cps;
            self.adj_first_click = self.first_click * 0.75;
        }

        if current_cps > self.first_click * 1.05 {
            self.first_click = current_cps;
            self.adj_first_click = self.first_click * 0.75;
        }

        if self.first_click > 0.0 && current_cps <= self.adj_first_click {
            return Some(self.game_over());
        }

        if current_cps > self.max_cps {
            self.max_cps = current_cps;
        }

        Some(GameEvent::Tick {
            seconds: self.seconds,
            total_clicks: self.total_clicks,
            current_cps,
        })
    }

    fn game_over(&mut self) -> GameEvent {
        self.stop();
        let elapsed = if self.mode == GameMode::Sprint {
            self.start_seconds - self.seconds
        } else {
            self.seconds
        };
        let avg_cps = if elapsed > 0.0 {
            self.total_clicks as f32 / elapsed
        } else {
            0.0
        };
        GameEvent::GameOver {
            taunt: self.get_taunt(),
            total_clicks: self.total_clicks,
            seconds: elapsed,
            avg_cps,
            max_cps: self.max_cps,
        }
    }

    fn get_taunt(&self) -> String {
        let total = self.total_clicks;
        if total > 350 {
            "You Might Be Playing too much MouseRacer!"
        } else if total > 251 {
            "All time high score! NOT!"
        } else if total > 201 {
            "Unstoppable!"
        } else if total > 151 {
            "That call stack has your name written all over it!"
        } else if total > 121 {
            "You Have Taken The Lead"
        } else if total > 101 {
            "you have achieved hyperclick!"
        } else if total > 50 {
            "You make Chuck Norris Proud!"
        } else {
            "you might consider a different hobby"
        }
        .to_string()
    }
}
